### PR TITLE
Support column aliases specified by `WITH` statements

### LINKED ISSUE #3716: Allow with statements to specify their columns alongside their expression names
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
PostgreSQL (and many others) allow `WITH` statements to take a list of column aliases right after the CTE name.

```sql
WITH numbers(a, b, c) AS (
    SELECT 1, 2, 3
)
SELECT * FROM numbers;
```

DataFusion:
```
+----------+----------+----------+
| Int64(1) | Int64(2) | Int64(3) |
+----------+----------+----------+
| 1        | 2        | 3        |
+----------+----------+----------+
```

PostgreSQL:
```
 a | b | c 
---+---+---
 1 | 2 | 3
```

**Describe the solution you'd like**
DataFusion could also support this syntax, considering sqlparser-rs already parses and relays all the relevant information.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by isidentical**
# Which issue does this PR close?

Closes #3716.

 # Rationale for this change

Adds support for processing the column aliases already parsed by sqlparser-rs during SQL planning stage.

# What changes are included in this PR?

Normally we already had some relevant logic processing these entries for relations, now it is in a separate function where both the CTE planner and the relation planner can use.

# Are there any user-facing changes?

This is a fix/feture.

### CODE REVIEW

File: datafusion/sql/src/planner.rs [hunk 1]
```diff
@@ -5013,6 +5025,67 @@ mod tests {
...
 
+    #[test]
+    fn cte_with_no_column_names() {
+        let sql = "WITH \
+        numbers AS ( \
+            SELECT 1 as a, 2 as b, 3 as c \
+        ) \
+        SELECT * FROM numbers;";
+
+        let expected = "Projection: #numbers.a, #numbers.b, #numbers.c\
```
Comment (andygrove) [inline datafusion/sql/src/planner.rs:5036]: Could you merge lastest from master - we no longer prefix column names with `#`

Comment (isidentical) [reply datafusion/sql/src/planner.rs]: Should be done now, thanks for reminding!

File: datafusion/sql/src/planner.rs [hunk 2]
```diff
@@ -785,33 +790,40 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                    .build()?)
-            }
+            self.apply_table_alias(plan, alias)
         } else {
             Ok(plan)
         }
     }
 
+    /// Apply the given TableAlias to the top-level projection.
+    fn apply_table_alias(
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:800]: 👌

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -5046,6 +5058,67 @@ mod tests {
...
+    #[test]
+    fn cte_with_column_aliases_precedence() {
+        // The end result should always be what CTE specification says
+        let sql = "WITH \
+        numbers(a, b, c) AS ( \
+            SELECT 1 as x, 2 as y, 3 as z \
+        ) \
+        SELECT * FROM numbers;";
+
+        let expected = "Projection: numbers.a, numbers.b, numbers.c\
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:5101]: 👍

**Review by alamb — APPROVED** [review]
This is beautiful @isidentical  -- thank you

**Pull request merged**